### PR TITLE
Relay maintainer block pull

### PR DESCRIPTION
**Pull request by tomaszslabon**
Depends on: #767 

This pull request contains functionalities responsible for pulling block headers from the Bitcoin network and putting them on the queue to be forwarded to the ETH network.

Work done so far:
- basic structure for the pulling loop
- added `GetHeaderByDigest`
- added finding best block to start pulling loop
- implemented pulling loop
- unit tests for pulling header loop

### CODE REVIEW

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/forwarder.go:99]: I think it may happen that:
1. We register new best header with digest X and height Y on the host chain
2. We crash soon due to an error
3. A reorg happens on BTC chain and height Y now corresponds to a digest Z 
4. We recover from the crash and initialize the pulling loop so we need to make sure what happened in the meantime on BTC chain

**Review by lukasz-zimnoch — CHANGES_REQUESTED** [review]
First pass. Probably more to come later.

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go]: I added some comments explaining why we do so.

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go:76]: Good idea.

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go]: Right

File: relay/pkg/btc/remote/remote.go
Comment (tomaszslabon) [reply relay/pkg/btc/remote/remote.go:159]: I didn't find any cleaner btc api method for returning block height based on block hash.

I see one alternative:
Instead of calling `GetBlockHeader` and `GetBlockHeaderVerbose` we could just call `GetBlockHeaderVerbose` as it contains all the information we need.
But `GetBlockHeaderVerbose` returns data in not very useful format (for example hashes are returned as strings).

We would have to construct `wire.BlockHeader` structure (which is needed for serialization) from the object returned from `GetBlockHeaderVerbose`  which would probably add more code.

File: relay/pkg/node/node.go
Comment (tomaszslabon) [reply relay/pkg/node/node.go]: I agree,  using quit channel was a really messy solution.

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go]: Yes, we can simply use pointers.

File: relay/pkg/btc/remote/remote.go
Comment (lukasz-zimnoch) [reply relay/pkg/btc/remote/remote.go:159]: Ok fine. I think we can leave it as is.

File: relay/pkg/btc/btc.go
Comment (lukasz-zimnoch) [inline relay/pkg/btc/btc.go:65]: Just a style note: It's a general Go convention that we use short abbreviations for receiver names. In those cases it should be just `(h *Header)`. If the struct name is more complicated, for example, `BitcoinBlockHash`, we use first letters of all parts: `(bbh *BitcoinBlockHash)`

File: relay/pkg/btc/btc.go
Comment (lukasz-zimnoch) [inline relay/pkg/btc/btc.go:21]: `btc/local` chain doesn't implement this method and this cause failure of the CI job

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go:103]: I created `pull.go` file and moved there `findBestBlock` and also functions `pullHeaderFromBtcNetwork` and `pushHeaderToQueue` that I created.

As for extracting more logic from `default`, it is hard to extract anything more, as there is a loop there and the logic in that loop uses variables (`nextHeaderHeight` and `lastAdded`) which are outside the scope of the loop.

I would have to pass them to the extracted function.

File: relay/pkg/block/pull.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull.go:5]: Let's call it `pullHeaderFromBtcChain` to be consistent with what we have on the push side.

File: relay/pkg/block/pull.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull.go:13]: `findBestBlock` -> `findBestHeader`?

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:103]: What about making `nextHeaderHeight` and `lastAdded` private fields of the `Forwarder` struct? We already use `processedHeaders` on the push side so we may introduce:
- `nextPulledHeaderHeight`
- `lastPulledHeader`
This may open more possibilities to take processing logic out of the pulling loop. What do you think?

Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go:103]: I created  `pullNextHeader` and moved it to `pull.go`.
I could move there even more logic from `default` there, but I think I will be easier to think about the pulling loop if we leave the sleeping part on top, so it is clearly visible.

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go]: That makes sense. Thanks.

File: relay/pkg/block/pull.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull.go:38]: There is no sense to make a setter for `nextPullHeaderHeight` in that case. We're not doing any validation or complex logic here. We can set the `nextPullHeaderHeight` field directly from calling code.

File: relay/pkg/block/forwarder.go
Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:103]: I get your point about the sleeping part but I think that sleep is strictly related to the "pull from BTC chain" part not the entire pull loop. My proposition of how the `default` case should look:
```
		default:
                        logger.Infof("pulling new header from BTC chain")

			header, err := f.pullHeaderFromBtcChain(ctx)
			if err != nil {
				f.errChan <- fmt.Errorf("could not pull header: [%v]", err)
				return
			}

                        logger.Infof("pushing new header to the queue")

			f.pushHeaderToQueue(header)
```
And that's it. Regarding methods used here look at the next comment.

File: relay/pkg/block/pull.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull.go:36]: Regarding those guys, the problem is as follows:
- `pullNextHeader` does too much. It's name suggests it only pulls next header but it actually pushes it to the queue and sets some internal forwarder state.
- `pullHeaderFromBtcChain` and `pushHeaderToQueue` don't do anything deserving a separate method

I'd rather go that way:
```
func (f *Forwarder) pullHeaderFromBtcChain(
	ctx context.Context,
) (*btc.Header, error) {
	for {
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}

		chainHeight, err := f.btcChain.GetBlockCount()
		if err != nil {
			return nil, fmt.Errorf("could not get block count [%v]", err)
		}

		// Check if there are more headers to pull or we are above the chain's
		// tip and need to sleep until the chain adds more headers
		if f.nextPullHeaderHeight <= chainHeight {
			nextHeader, err := f.btcChain.GetHeaderByHeight(f.nextPullHeaderHeight)
			if err != nil {
				return nil, fmt.Errorf(
					"could not get header by height at [%d]: [%v]",
					f.nextPullHeaderHeight,
					err,
				)
			}

			// TODO: Consider just comparing hashes - should be enough
			//  and check if nextHeader and lastPulledHeader header can ever
			//  be not equal
			if !nextHeader.Equals(f.lastPulledHeader) {
				return nextHeader, nil
			}
		}

		select {
		case <-time.After(forwarderPullingSleepTime):
		case <-ctx.Done():
		}
	}
}

func (f *Forwarder) pushHeaderToQueue(header *btc.Header) {
	f.headersQueue <- header
	f.lastPulledHeader = header
	f.nextPullHeaderHeight++
}
```
That way we solve all the problems and we obtain a clean pulling loop. All pull logic lives in `pull.go` and reading from BTC chain is separated from pushing to the queue. What do you think?

**Review by lukasz-zimnoch — CHANGES_REQUESTED** [review]

File: relay/pkg/block/pull.go
Comment (tomaszslabon) [reply relay/pkg/block/pull.go]: I'm ok with that.

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go]: I left a TODO and added "Go rewrite: Proper support of BTC reorgs" task in code.

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go]: Excellent, thanks!

**Comment by lukasz-zimnoch** [comment]
No more comments regarding the code. Waiting for unit tests now!

File: relay/pkg/block/forwarder_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/forwarder_test.go:33]: Can we shorten this sleep? Maybe `100ms` will be enough?

File: relay/pkg/block/forwarder_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/forwarder_test.go:41]: I'd add a short sleep (`100ms`?) after context cancel. This is needed to make sure nothing is added on the next pulling loop iteration which will occur after the one which has been ended by context cancellation.

File: relay/pkg/block/forwarder_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/forwarder_test.go:50]: We usually use the `expected & actual` pattern:
```
        expectedQueueLength := 0
	actualQueueLength := len(forwarder.headersQueue)
	if expectedQueueLength != actualQueueLength {
		t.Errorf(
			"unexpected headers queue length:\n"+
				"expected: [%v]\n"+
				"actual:   [%v]\n",
			expectedQueueLength,
			actualQueueLength,
		)
	}
```

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:63]: Hint: We can also use `%+v` placeholder to print struct fields along with their names:
```
"unexpected header in queue:\n"+
        "expected: [%+v]\n"+
        "actual:   [%+v]\n",
    expectedHeader,
    actualHeader,
```

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:101]: How do we know we are actually not waiting here? I think we need some kind of timeout mechanism here to make sure headers are returned without delay.

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:87]: Can we make that time lower? How about `1s` or even `500ms`?

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:132]: If we shorten the loop sleeping time, we can use a smaller value here as well.

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:134]: Please add space between `returned` and `[%v]`

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:143]: Maybe `header timeout has been exceeded`?

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:142]: If we shorten the loop sleeping time, we can use a smaller value here as well.

File: relay/pkg/block/pull_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/pull_test.go:201]: `...DoesNot...`

File: relay/pkg/block/push_test.go
Comment (lukasz-zimnoch) [inline relay/pkg/block/push_test.go:22]: Do we need to set `forwarderPullingSleepTime`? Is this test failing without that?

Comment (tomaszslabon) [reply relay/pkg/block/push_test.go:22]: It's not needed here.

File: relay/pkg/block/pull_test.go
Comment (tomaszslabon) [reply relay/pkg/block/pull_test.go:143]: Shortened it to 300ms.

File: relay/pkg/block/pull_test.go
Comment (tomaszslabon) [reply relay/pkg/block/pull_test.go]: I changed the test a bit and ran these two `pullHeaderFromBtcChain` calls on a separate goroutine and added a timeout.

**Review by lukasz-zimnoch — APPROVED** [review]
LGTM! 🚀 I'll merge once #767 lands on master!

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:32]: Can we add one or two sentences explaining why it needs/wants to rest?

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:33]: Why do we name it `default*` given that it is not configurable and the rest of the constants defined in this section are not prefixed with `default`?

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [inline relay/pkg/block/forwarder.go:85]: Is it block pull or header pull?

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:85]: It originates from the approach taken in #767. I think we could leave as is and I will refactor that according to https://github.com/keep-network/tbtc/pull/767#discussion_r603913978

Comment (pdyraga) [reply relay/pkg/block/forwarder.go:85]: Yeah, agreed. I wrote this comment before I realized what is happening and the discussion in https://github.com/keep-network/tbtc/pull/767#discussion_r603913978 😆

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:32]: For example, we could add the following sentence:
> The forwarder waits for this time before it tries to fetch a new tip from the Bitcoin blockchain, giving it some time to mine new blocks.

Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go:32]: Perhaps:
"Default duration for which the forwarder should rest after reaching the tip of Bitcoin blockchain as there are no more headers to pull"
?

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go:33]: `forwarderPullingSleepTime` inside `Forwarder` is configurable - in tests we use a shorter value, while in real implementation we use `defaultForwarderPullingSleepTime`.

Perhaps we could name this constant `forwarderPullingSleepTime`, while the field inside `Forwarder` could be named `poolingSleepTime`?

Comment (lukasz-zimnoch) [reply relay/pkg/block/forwarder.go:33]: Maybe `pullingSleepTime`?

Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go:33]: Yes :)

Comment (pdyraga) [reply relay/pkg/block/forwarder.go:33]: I see. Do we plan to make this value configurable in the config file? Because if so, prefix makes sense. Otherwise, I'd name it `forwarderPullingSleepTime` in both cases, we can use the same name.

File: relay/pkg/block/forwarder.go
Comment (pdyraga) [reply relay/pkg/block/forwarder.go:32]: FWIW, I do not consider this or any other comment I left here blocking. Please feel free to merge @lukasz-zimnoch and we can address them in any other PR we have in the queue or a new one.

File: relay/pkg/block/forwarder.go
Comment (tomaszslabon) [reply relay/pkg/block/forwarder.go:33]: We do not plan to make it configurable via config file at this moment.
I'm ok with `forwarderPullingSleepTime` both in Forwarder and in const section.

Comment (pdyraga) [reply relay/pkg/block/forwarder.go:33]: `pullingSleepTime` is even better!

**Pull request merged**
